accepted/2021/User-Experience/Apple-Silicon-UX.md

```diff
@@ -0,0 +1,254 @@
+
+Overall the experiment demonstrated a highly workable user experience.
+
+- The experiment represents a viable design alternative. It did not support side-by-side install of multiple architectures of the same runtime version.
+
+- The implementation was incomplete. It did not automate the build process or deal with installer issues
+
+### Dotnet Publish Experience Inconsistent
+
+#### Signing
+
+- dotnet publish for Xamarin workloads handles signing automatically.
+- dotnet publish for CoreClr console apps do not handle signing automatically. The developer would need to explicitly sign after publishing for Apple Silicon.  The signing command is relatively simple `codesign -s <signature> --entitlements <app-entitlements-path> <app-path>`.
+- There is no message to the customer that the published app needs to be signed
+
+#### Dotnet Publish as Universal Binaries
+
+- This scenario is also supported for Xamarin iOS workloads. This is a slightly different scenario as Xamarin iOS apps are native binaries (not JIT binaries)
+- CoreCLR runtime apps have no mechanism to publish universal binaries.
+
+### Dotnet-SOS and the SOS plugin Issues
+
+The `dotnet-sos` tool installs the `SOS` diagnostic plugin into `lldb`. Since `lldb` is a universal binary, it expects its plugins to also be universal binaries. With the single architecture design, when `SOS` is installed, `lldb` will complain about not being able to load the plugin when the architecture mismatches.
+
+#### Universal Binary Experiment
+
+Making the Apple Silcon SOS be a universal binary to allowed SOS to work on `x86_64` and `arm64` processes.
+
+## User Experience Designs
+
+### Mixed Architecture Side By Side Installs
+
+This is controversial. This is _uncommitted/unplanned_. Ongoing discussions will likely affect this proposal.
+
+For CoreCLR runtime, the proposal would be to:
+
+- Make `dotnet` & `hostfxr` to be universal binaries
+- Support automatic architecture selection
+  - Modify `hostfxr` to assume .NET 5 and earlier require x64 architecture.
+  - Modify `dotnet` to restart in the correct architecture on an architecture mismatch.
+- Make the CoreCLR runtime effectively a Universal Binaries
+  - Create a single distribution for both macOS architectures.
+  - Make the runtime native binaries to be universal binaries
+  - Modify `hostpolicy` to probe for trusted platform assemblies based on the current RID. This is designed to architecture specific ready-to-run managed assemblies. This RID specific path would be probed before probing the common any platform assembly path.
+  - Modify the runtime probing to be consistent with the new `hostpolicy` probing.
+- Modify the build system to build the Universal Binaries
+- Modify installers
+  - Modify .NET 5 installers to allow installation without breaking universal `dotnet`.
+  - Modify .NET 6 installer to overwrite the x64 `dotnet`
+
+#### Open Questions
```

Comment:
My personal preference:
- Choose a different location for installation of `x64` .NET 6+ runtime - something like `dotnet_x64`. The existence of this is transient and in some not so distant future release will go away
- The primary install location for .NET 6+ will be `arm64` only (for the most part)
  - Keep the muxer/hostfxr as universal binaries in the main install location
  - Add some logic to also look for runtimes/SDKs in the `dotnet_x64` location - the universal muxer/hostfxr would use this to get the right runtime if asked to run `x64` app
- The primary location would be the only one registered (in config file and so on) and it would be the only one on "path".

Reply:
After reading this document, the above statement by @vitek-karas would be my preference for a solution.

Reply:
> Choose a different location for installation of x64 .NET 6+ runtime - something like dotnet_x64.

Are you proposing we backport this change to the 3.1/5.0 branches.

I was thinking we could only do this for the .NET 6+ branches.

Reply:
I guess the advantage of @vitek-karas's approach would be that we just need to teach the installer to alter the install path for x64 on Apple Silicon.

That seems easier that fixing it overwriting the dotnet and hostfxr universal binaries.

That said, I am leanig to @vitek's approach too.

Reply:
If we never install .NET 6+ x64 into global location (we simply don't produce the installers for this) then the `dotnet_x64` approach is not much different from keeping everything in `dotnet`. Obviously this is ignoring the potential problems with overwriting the `dotnet` (note that `hostfxr` doesn't have a problem since it's installed into a version specific directory, so the older installers will never overwrite the newer one).

The only true advantage of having `dotnet_x64` is that it would allow to install both .NET 6 x64 and .NET 6 arm64 side by side.

Reply:
>The only true advantage of having dotnet_x64 is that it would allow to install both .NET 6 x64 and .NET 6 arm64 side by side.

It would also give us a pattern to follow - which is important here. Having a process that we can fallback on is important for our customers and ourselves when we rediscover this problem in 18 months.